### ISSUE TITLE
Metro-Police

### ISSUE THREAD

**Issue by BEENNath58** [narzoul/DDrawCompat#373]
The game can be downloaded from old-games.ru. I recommend patching it to 1.6 version from patches-scrolls de because without it the game crashes.

On modern Windows the game needs FaultTolerantHeap to run well.

If you have a DRM related issue, create a folder named "LASERLOK" and put an empty file "LASERLOK.IN" and the game should run.

The color issue is from the use of palleted textures. It is easily solvable by enabling it in the .ini, but the screenshots were taken before I set it, so please ignore that part.

1. The game sets **transparency on the ladders** as expected:
![Screenshot (2)](https://github.com/user-attachments/assets/6ddeca5a-08d4-4670-8542-468668a000a7)

However, with palleted texture support in DDrawCompat and with dgVoodoo2:
![Screenshot (1)](https://github.com/user-attachments/assets/16c30f14-2a4d-4e8f-9353-05f984a053e3)

On an emulated Voodoo2:
![voodoo2](https://github.com/user-attachments/assets/0e6ccaeb-1bd4-4810-a94e-95ba3e67c821)

2. The loading screen text should "scroll", however it just comes up as a text. On dgVoodoo2 it does fine:
![loading](https://github.com/user-attachments/assets/434897d2-b139-47c1-80b5-f014e3900846)

3. The game **water is darker** than normal:
![darkmp](https://github.com/user-attachments/assets/317aad9a-66d5-48cb-8935-b743f37175b4)

For testing: you can use the savefiles from [here](https://sourceforge.net/p/dxwnd/discussion/general/thread/aeaf5e01c1/5eb9/70a0/attachment/savegame.7z).
The "texture2" save should get you to the water bug in point 3, and the "423432423" save should get you to the transparency bug in point 1.

Thank you

**Comment by narzoul** [comment]
The game still consistently crashes while loading a saved game or starting a new game, at about a quarter of the progress bar. That's with FaultTolerantHeap. Without it, it crashes before the progress bar even starts.

**Comment by BEENNath58** [comment]
I randomly decided to hook the game to DxWnd, with DirectX7 version hook and Handle Exceptions, and the game runs in mission. Works without FaultTolerantHeap too.

**Comment by narzoul** [comment]
I figured it out based on the discussion on DxWnd forums. It was an issue with the wrong directory names in the RIP. I just replaced the whole Scene directory with the one from the CD install (which I couldn't get working with the empty laserlok.in file), and now it works.

> The game sets transparency on the ladders as expected:

It sets the wrong color key when palettized textures are used. Those black pixels have a palette index of 0x7F but the game sets the color key to 0xFF. Both map to the same black color in the palette, so I suppose it only looks "correct" on some buggy drivers that match the color key with the looked-up color rather than the index. It also looks wrong with the Direct3D software RGB renderer (you can try it on PCem with a non-3D capable GPU), so I don't want to touch it. It should be patched in the game.

> The loading screen text should "scroll", however it just comes up as a text. On dgVoodoo2 it does fine:

It's a matter of performance. The game draws each letter one by one with BltFast as quickly as it can. I recorded it with OBS at 60 FPS on a 144Hz monitor. With DDrawCompat, the whole text is rendered in 1-2 frames (~5 with debug logs enabled). With dgVoodoo it takes ~12.

> The game water is darker than normal:

Darker than what? To me, it looks about the same as on Voodoo 2 in that particular save.

**Comment by BEENNath58** [comment]
> Darker than what? To me, it looks about the same as on Voodoo 2 in that particular save.

Here, it looks:
![Screenshot from 2024-10-08 04-06-05](https://github.com/user-attachments/assets/9c8e923c-b289-4d92-9f50-5fb24475878c)

**Comment by narzoul** [comment]
I don't get it. That still looks pretty dark. Where is the "non-dark" version? And with what hardware or wrapper?

**Comment by BEENNath58** [comment]
> I don't get it. That still looks pretty dark. Where is the "non-dark" version? And with what hardware or wrapper?

I am talking about the water. By dark, I meant to say it's more bluer, and more specifically, non transparent too. 

Its a Voodoo 3

**Comment by narzoul** [comment]
I think the non-transparent version matches the Voodoo 2 and the transparent version matches the Voodoo 3. Of course they look different, since the transparent version has colors mixed with the bottom of the pool...
As for why it's transparent in one case and not the other, I have no idea. It must be an engine bug, probably it skips alpha blending when it shouldn't or vice versa. Maybe related to all the memory faults. With RGB565 texture format it's always transparent, but with palettized textures it depends on who-knows-what.

**Comment by narzoul** [comment]
Ok, I think I figured out the transparency issue thanks to a hint from the DX6 SDK:

> D3DTBLEND_MODULATE 
Modulate texture-blending mode is supported. In this mode, the RGB values of the texture are multiplied with the RGB values that would have been used with no texturing. Any alpha values in the texture replace the alpha values in the colors that would have been used with no texturing; **if the texture does not contain an alpha component, alpha values at the vertices in the source are interpolated between vertices.**

DDrawCompat always emulates palettized textures via the A8R8G8B8 format, so the texture itself always has alpha values. Therefore, the alpha was taken from the palette rather than the vertices. The only thing I don't understand yet is why the transparency still worked in the other save.

I replaced it with X8R8G8B8, which breaks the palettized texture support in FF8 instead. But it turns out that it was possible to fix by removing the D3DPTEXTURECAPS_ALPHAPALETTE capability. In this case, the game falls back to using color keying instead of palettes with alpha.

I added these changes, along with the ladder "fix", so that the Voodoo 3 behavior can be reproduced more accurately in this game:
[ddraw.zip](https://github.com/user-attachments/files/17292751/ddraw.zip) ([diff.txt](https://github.com/user-attachments/files/17292754/diff.txt) compared to v0.5.4)

I don't know if this breaks any other palettized texture games though. If it turns out that no other games need D3DPTEXTURECAPS_ALPHAPALETTE support, and that this alternative method of color key matching works fine everywhere, then I suppose it could stay this way by default in the next release. I really don't want to add more unnecessary config options.
